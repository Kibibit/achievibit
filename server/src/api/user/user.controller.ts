import {
  Controller,
  Logger,
  NotFoundException,
  Param,
  UseFilters
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { GetAll, GetOne } from '@kb-decorators';
import { KbValidationExceptionFilter } from '@kb-filters';
import { User } from '@kb-models';

import { UserService } from './user.service';

@Controller('api/user')
@ApiTags('user')
@UseFilters(new KbValidationExceptionFilter())
export class UserController {
  private readonly logger = new Logger(UserController.name);

  constructor(private readonly userService: UserService) {}

  @GetAll(User)
  async getAllUsers() {
    const users = await this.userService.findAllUsers();

    return users;
  }

  @GetOne(User, ':username')
  async getUser(@Param('username') username: string) {
    const user = await this.userService.findByUsername(username);

    if (!user) {
      throw new NotFoundException(`User with username ${ username } not found`);
    }

    // will show secret fields as well!
    this.logger.log('Full User');
    // will log only public fields!
    this.logger.log(user);
    // DANGER! WILL LOG EVERYTHING!
    // console.log(user);

    // will only include exposed fields
    return user;
  }
}
