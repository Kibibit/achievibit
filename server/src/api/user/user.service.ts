import { ReturnModelType } from '@typegoose/typegoose';

import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';

import { BaseService } from '@kb-abstracts';
import { User } from '@kb-models';

@Injectable()
export class UserService extends BaseService<User> {
  constructor(
    @InjectModel(User.modelName)
    private readonly userModel: ReturnModelType<typeof User>
  ) {
    super(userModel, User);
  }

  async findAllUsers(): Promise<User[]> {
    const dbUsers = await this.findAll().exec();

    return dbUsers.map((user) => new User(user.toObject()));
  }

  async findByUsername(username: string): Promise<User> {
    const dbUser = await this.findOne({ username }).exec();

    if (!dbUser) {
      return;
    }

    return new User(dbUser.toObject());
  }
}
