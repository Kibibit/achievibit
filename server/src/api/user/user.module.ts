import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';

import { User } from '@kb-models';

import { UserController } from './user.controller';
import { UserService } from './user.service';

@Module({
  imports: [
    MongooseModule.forFeature([
      { name: User.modelName, schema: User.schema }
    ])
  ],
  providers: [UserService],
  controllers: [UserController]
})
export class UserModule {}
