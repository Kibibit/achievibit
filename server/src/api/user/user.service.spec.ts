import { MongooseModule } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';

import {
  closeInMemoryDatabaseConnection,
  createInMemoryDatabaseModule,
  DtoMockGenerator
} from '@kb-dev-tools';
import { User } from '@kb-models';

import { UserService } from './user.service';

describe('UserService', () => {
  let service: UserService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        createInMemoryDatabaseModule(),
        MongooseModule.forFeature([{
          name: User.modelName,
          schema: User.schema
        }])
      ],
      providers: [UserService]
    }).compile();

    service = module.get<UserService>(UserService);
  });

  afterEach(async () => {
    await closeInMemoryDatabaseConnection();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be able to get all', async () => {
    const user1 = DtoMockGenerator.user();
    const user2 = DtoMockGenerator.user();

    const createdUser1 = await service.create(user1);
    const createdUser2 = await service.create(user2);

    const foundUsers = await service.findAll();
    const foundUsersParsed = foundUsers
      .map((user) => new User(user.toObject()));

    expect(foundUsersParsed).toContainEqual(createdUser1);
    expect(foundUsersParsed).toContainEqual(createdUser2);
    expect(foundUsersParsed.length).toBeGreaterThanOrEqual(2);
  });

  it('should be able to get all users', async () => {
    const user1 = DtoMockGenerator.user();
    const user2 = DtoMockGenerator.user();

    const createdUser1 = await service.create(user1);
    const createdUser2 = await service.create(user2);

    const foundUsers = await service.findAllUsers();

    expect(foundUsers).toContainEqual(createdUser1);
    expect(foundUsers).toContainEqual(createdUser2);
    expect(foundUsers.length).toBeGreaterThanOrEqual(2);
  });

  it('should be able to get a user by username', async () => {
    const user = DtoMockGenerator.user();

    const createdUser = await service.create(user);

    const foundUser = await service.findByUsername(user.username);

    expect(foundUser).toEqual(createdUser);
  });

  it('should return null if user does not exist', async () => {
    const foundUser = await service.findByUsername('username');

    expect(foundUser).toBeUndefined();
  });

  it('should NOT allow creating 2 users with same username', async () => {
    const user1 = new User({
      ...DtoMockGenerator.user(),
      username: 'test-user'
    });
    const user2 = new User({
      ...DtoMockGenerator.user(),
      username: 'test-user'
    });

    const createdUser = await service.create(user1);
    const regex = /([\d\w-]*?.users)/g;
    let creationError: string;
    await service.create(user2)
      .catch((err) => creationError = err.message);
    creationError = creationError.replace(regex, 'users');

    expect(creationError).toMatchSnapshot();

    const allUsers = await service.findAllUsers();
    expect(allUsers.length).toBe(1);
    expect(allUsers[0]).toEqual(expect.objectContaining(createdUser));
    expect(allUsers[0]._id).toBeDefined()
    expect(allUsers[0]._id).toEqual(createdUser._id);
  });
});
