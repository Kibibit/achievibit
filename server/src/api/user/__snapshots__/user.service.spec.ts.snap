// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService should NOT allow creating 2 users with same username 1`] = `"E11000 duplicate key error collection: users index: username_1 dup key: { username: \\"test-user\\" }"`;
